Produce 'invisible' squiggles for 'Remove Unnecessary Usings' and 'Simpl...

shyamnamboodiripad opened the pull request:
...ify Type Name'.

Produce invisible / suggestion squiggles (which will in turn display Quick Info on mouse hover) for the hidden diagnostics that we report for 'Remove Unnecessary Usings' and 'Simplify Type Name'. The presence of Quick Info pane for such squiggles allows allows platform to display Light Bulb for the corresponding fixes. Per their current design platform can only display light bulb if Quick Info pane is present.

Here's some screenshots that show how the quick info + light bulb will look -
![1](https://cloud.githubusercontent.com/assets/10579684/6643251/20683476-c968-11e4-842a-4bb62438d384.PNG)

If we have multiple issues on the line -
![2](https://cloud.githubusercontent.com/assets/10579684/6643272/44bef90e-c968-11e4-9ad3-676d9580336f.PNG)

shyamnamboodiripad commented:
@srivatsn @heejaechang @mavasani @tmeschter @jmarolf - could you please take a look?

FYI @balajikris

shyamnamboodiripad commented:
@dotnet-bot test this please

srivatsn commented on src/Features/Core/Diagnostics/Analyzers/RemoveUnnecessaryImportsDiagnosticAnalyzerBase.cs:
```diff
@@ -16,16 +16,6 @@ internal abstract class RemoveUnnecessaryImportsDiagnosticAnalyzerBase : Diagnos
         // NOTE: This is a trigger diagnostic, which doesn't show up in the ruleset editor and hence doesn't need a conventional IDE Diagnostic ID string.
         internal const string DiagnosticFixableId = "RemoveUnnecessaryImportsFixable";
 
-        private static LocalizableString s_localizableMessageAndTitle = new LocalizableResourceString(nameof(WorkspacesResources.RemoveUnnecessaryImportsOrUsings), WorkspacesResources.ResourceManager, typeof(WorkspacesResources));
```
It would be nice to keep the descriptor here so that both of the descriptors are together instead of being spread across three files. Instead of the abstract GetClassificationIdDescriptor, you could simply have an abstract GetMessage(). That way if we make changes to the descriptor we can make it in one place.

shyamnamboodiripad replied:
Sure will change that.

heejaechang replied:
+1

VladimirReshetnikov commented on src/EditorFeatures/Core/Implementation/Diagnostics/DiagnosticsSquiggleTaggerProvider.TagSource.cs:
```diff
@@ -39,7 +39,8 @@ protected override int MinimumLength
                 return
-                    (diagnostic.Severity == DiagnosticSeverity.Warning || diagnostic.Severity == DiagnosticSeverity.Error) &&
+                    (diagnostic.Severity == DiagnosticSeverity.Warning || diagnostic.Severity == DiagnosticSeverity.Error ||
+                    (diagnostic.Severity == DiagnosticSeverity.Hidden && diagnostic.CustomTags.Contains((c) => c == WellKnownDiagnosticTags.Unnecessary))) &&
```
Can we add an overload of `Contains` (or is there already?) that takes the value searched for rather than a predicate. I guess it could save some allocations.

shyamnamboodiripad replied:
Thanks - I couldn't find one, but I will add.

heejaechang replied:
can we make the expression a bit cleaner?

srivatsn commented:
Jenkins is failing because PredefinedErrorTypeNames.Suggestion is a new value that's not in the CTP6 build that Jenkins uses. I think we should have a conditional constant in RoslynLight.sln and then ifdef out this statement when that constant is not defined.

heejaechang commented:
:+1:

shyamnamboodiripad commented:
Introducing solution level conditional constant in a way that works smoothly for both command line and VS builds of this solution is complicated. For now, I've removed the projects that depend on the new API from RoslynLight.sln to fix Jenkins. 

I've addressed the remaining pieces of feedback above. Waiting for Jenkins build to succeed. I will go ahead and merge this once that succeeds.

shyamnamboodiripad commented:
@dotnet-bot test this please

VladimirReshetnikov commented on src/EditorFeatures/Test/Squiggles/AbstractSquiggleProducerTests.cs:
```diff
@@ -29,8 +29,18 @@ protected static IEnumerable<ITagSpan<IErrorTag>> GetErrorSpans(TestWorkspace wo
-            var analyzerService = new DiagnosticAnalyzerService(compilerAnalyzersMap);
+
+            DiagnosticAnalyzerService analyzerService = null;
+            if (analyzerMap == null || analyzerMap.Count == 0)
```
You could write `analyzerMap?.Count > 0` and flip then/else branches.
